### PR TITLE
Extend `istioctl tag set default` to control validation

### PR DESCRIPTION
**Pull request by Monkeyanator**
Implements [https://docs.google.com/document/d/1P1mNlrJOs-KHVqcKTrFLrvCUiLgyZ-Vlf7ezaZNvD6g/edit?usp=sharing](https://docs.google.com/document/d/1P1mNlrJOs-KHVqcKTrFLrvCUiLgyZ-Vlf7ezaZNvD6g/edit?usp=sharing)

Makes it so that running `istioctl tag set default --revision` moves which revision handles validation.

- [x] Disable validation by default on revisioned webhooks
- [x] Ensure that each revision patches validating webhooks with matching istio.io/rev
- [x] Add template for default validating webhook configuration
- [x] Create logic in istioctl tag to create and generate the default validator
- [ ] Istioctl tag delete default should remove the default validator
- [x] Make it possible to point default tag at pre-1.11 revisions
- [ ] Write integration test for istioctl tag set validation changes
- [x] Add logic to installer to make an Istio installation default if there is no existing default
- [x] Provide a Helm-centric way to switch default revision
- [x] Purge deprecated VWH when the default is set

Requires the following documentation changes:

1. For Helm install, when installing base, use the `--defaultRevision=<rev>` option.
2. Document the process of installing a new revision, evaluating it, and switching the default revision with `istioctl tag set default --revision <rev>`.
3. Document default revision semantics including (1) that it handles injection for `istio-injection=enabled` and (2) it handles resource validation for the entire cluster.

[ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: istioctl/pkg/tag/templates/validatingwebhook.yaml
Comment (howardjohn) [inline istioctl/pkg/tag/templates/validatingwebhook.yaml:1]: Just curious why we have a copy under istioctl for this, but mutating webhook is using the chart's version? Is this unique from the chart version?

File: istioctl/pkg/tag/templates/validatingwebhook.yaml
Comment (howardjohn) [inline istioctl/pkg/tag/templates/validatingwebhook.yaml:34]: This list is not complete, and i see its not in manifests/ either. So that is a bug in current master branch I think

File: istioctl/pkg/tag/templates/validatingwebhook.yaml
Comment (Monkeyanator) [reply istioctl/pkg/tag/templates/validatingwebhook.yaml:1]: Mutating webhooks can move here too since we decided being able to set tags through IstioOperator is not a priority.

File: istioctl/pkg/tag/templates/validatingwebhook.yaml
Comment (Monkeyanator) [reply istioctl/pkg/tag/templates/validatingwebhook.yaml:34]: Yup I noticed that was missing telem API group.

File: istioctl/pkg/tag/util.go
Comment (Monkeyanator) [inline istioctl/pkg/tag/util.go:142]: This tries to make it possible to switch back to a non-revisioned install after making a revisioned install the default, a bit gross

File: istioctl/pkg/tag/templates/validatingwebhook.yaml
Comment (Monkeyanator) [reply istioctl/pkg/tag/templates/validatingwebhook.yaml:1]: Will handle in future PR.

File: istioctl/pkg/tag/templates/validatingwebhook.yaml
Comment (costinm) [inline istioctl/pkg/tag/templates/validatingwebhook.yaml:3]: Would it be hard to move this to /manifests/charts/default/templates/ - in preparation for adding a Chart.yaml ? And also when executing the template, set the variables to match helm ( {{ .Values.tag }}, etc.

If it's too hard we can leave it for later - or just duplicate it. I have a chart with equivalent functionality in costinm/istiod/manifests/.. - I was planning to just leave it as a sample.

Alternatively, you could use the template from 'base', or add few extra {{ if }} to 
the base so it renders tags ( and helm install base --set tag=foo --set rev=bar applies just the webhooks, the rest is skipped ). This will keep things in sync.
( istioctl has code to render the chart, so may be even less code )

File: istioctl/pkg/tag/util.go
Comment (costinm) [reply istioctl/pkg/tag/util.go:141]: Right, but I can't think of a better option. For 1.11+ we should just not install this webhook, even for the 'revision=""' - so it will not be a long-term problem, once 1.10 is out of support we can remove this hack.

Make sure you document the behavior - and the fact that if user make changes to the 1.10 'unrevisioned', like enable access logs - it will revert your hack. 

It may be a case of just documenting a manual rollback procedure for 1.10 - and making sure in future we are fine.

**Review by costinm — COMMENTED** [review]
In general it seems fine and better than what is currently in master. 
But I suspect we may need to iterate a bit more.

**Review by nrjpoddar — COMMENTED** [review]
I want to make sure the [design doc](https://docs.google.com/document/d/1P1mNlrJOs-KHVqcKTrFLrvCUiLgyZ-Vlf7ezaZNvD6g/edit?ts=60e5ef35) is approved and the flow works for Helm along with istioctl before we merge this PR in.

**Comment by Monkeyanator** [comment]
> I want to make sure the [design doc](https://docs.google.com/document/d/1P1mNlrJOs-KHVqcKTrFLrvCUiLgyZ-Vlf7ezaZNvD6g/edit?ts=60e5ef35) is approved and the flow works for Helm along with istioctl before we merge this PR in.

Added a section for how this can work with Helm @nrjpoddar

File: istioctl/pkg/tag/util.go
Comment (howardjohn) [inline istioctl/pkg/tag/util.go:156]: Is it possible to just keep the MutatingWebhookConfiguration with empty list of Webhooks?

File: istioctl/pkg/tag/util.go
Comment (howardjohn) [inline istioctl/pkg/tag/util.go:91]: nit: GetWebhookTagName may be more accurate

File: operator/cmd/mesh/install.go
Comment (howardjohn) [inline operator/cmd/mesh/install.go:196]:
```suggestion
		cmd.Print("Making this installation the default for injection and validation.\n")
```

Comment (howardjohn) [reply operator/cmd/mesh/install.go:196]: or Println if they have it

File: istioctl/cmd/tag.go
Comment (howardjohn) [inline istioctl/cmd/tag.go:281]: Not sure I understand why this was inverted?

File: manifests/charts/istio-control/istio-discovery/templates/validatingwebhookconfiguration.yaml
Comment (howardjohn) [inline manifests/charts/istio-control/istio-discovery/templates/validatingwebhookconfiguration.yaml:53]: should we remove the istiod-istio-system one as well now?

File: operator/cmd/mesh/install.go
Comment (howardjohn) [inline operator/cmd/mesh/install.go:189]: I don't know how much of this is broken already, but `istioctl x unintsall` leads around all the webhooks (including validation) which ends up breaking

**Review by howardjohn — COMMENTED** [review]
overall looks good!

File: istioctl/pkg/tag/util.go
Comment (Monkeyanator) [reply istioctl/pkg/tag/util.go:156]: Yeah the reason I did it this way is we need to keep the injection URL around _somewhere_ so we can create new tags from the webhook. Right now the webhooks are the only place it's persisted.

File: istioctl/cmd/tag.go
Comment (Monkeyanator) [reply istioctl/cmd/tag.go:281]: When we automatically set the default revision in the installer, we need it to happen without user confirmation (like when we do `install -y` in test framework). This makes it so that we only do the duplicate injection analysis and confirmation if `skip-confirmation` is disabled.

File: manifests/charts/istio-control/istio-discovery/templates/validatingwebhookconfiguration.yaml
Comment (Monkeyanator) [reply manifests/charts/istio-control/istio-discovery/templates/validatingwebhookconfiguration.yaml:53]: I still don't think we can because of the resource pruning removing the old, active validator?

File: operator/cmd/mesh/install.go
Comment (Monkeyanator) [reply operator/cmd/mesh/install.go:189]: Hm yeah it seems like the webhooks get removed when doing `uninstall --purge` but not `uninstall --revision` (even though the `istio.io/rev` label is there)... I'll try and fix as part of this PR.

File: istioctl/pkg/tag/util.go
Comment (howardjohn) [reply istioctl/pkg/tag/util.go:156]: Ah, lame. Thats annoying. I don't know how often users are actually doing this but i was pretty confusing looking at `kubectl get mutatingwebhooks` and seeing a bunch, when the yaml shows disabled. In theory we could stick it in an annotation but I am not sure that is worth the complexity, since in some cases its not just a simple URL (port, cabundle, etc).

I guess the comment is accurate :slightly_smiling_face:

File: manifests/charts/istio-control/istio-discovery/templates/validatingwebhookconfiguration.yaml
Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/templates/validatingwebhookconfiguration.yaml:53]: Aren't we replacing it with a new active validator though? I guess not if we are not the default?

Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/templates/validatingwebhookconfiguration.yaml:53]: I think fixing the old validator is a must-do for 1.11

File: istioctl/pkg/tag/util.go
Comment (Monkeyanator) [reply istioctl/pkg/tag/util.go:156]: I have no doubt that someone, somewhere will eventually come to rely on `"istio.io/deactivated": "never-match"` for injection and we'll have to keep it around :smile:

File: istioctl/pkg/tag/templates/validatingwebhook.yaml
Comment (Monkeyanator) [reply istioctl/pkg/tag/templates/validatingwebhook.yaml]: Makes sense, moved the template to `manifests/charts/default` and the command now renders from there. There's also a `defaultRevision` option now on the base chart.

File: manifests/charts/istio-control/istio-discovery/templates/validatingwebhookconfiguration.yaml
Comment (Monkeyanator) [reply manifests/charts/istio-control/istio-discovery/templates/validatingwebhookconfiguration.yaml:53]: Just tested and it seems like removing the webhook doesn't actually purge it from old installations... should be clear to remove.

**Comment by howardjohn** [comment]
Should we explicitly prune it if/when we set a new default? That way users
won't have a broken webhook laying around?

File: manifests/charts/istio-control/istio-discovery/templates/validatingwebhookconfiguration.yaml
Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/templates/validatingwebhookconfiguration.yaml:53]: Should we explicitly purge it when we set the default? That way we don't get a stale webhook, which will either give noisy errors in Istiod at best, or block config at worst?

**Comment by Monkeyanator** [comment]
> Should we explicitly prune it if/when we set a new default?

@howardjohn the reason it's kept around here is so that users can switch back to it if things go wrong after the initial switch. Maybe we can just do away with the whole `DeactivateWebhook` hack and delete the webhook and document the caveat?

File: manifests/charts/istio-control/istio-discovery/templates/validatingwebhookconfiguration.yaml
Comment (Monkeyanator) [reply manifests/charts/istio-control/istio-discovery/templates/validatingwebhookconfiguration.yaml:53]: Made it so that (1) validating webhooks get removed on `istioctl uninstall --revision` and (2) we explicitly purge the old one when setting default.

File: pkg/webhooks/validation/controller/controller.go
Comment (Monkeyanator) [inline pkg/webhooks/validation/controller/controller.go:123]: had to make validation controller patch all webhooks with matching `istio.io/rev` since for Helm base install we need `istiod` to patch in the cert

File: operator/cmd/mesh/install.go
Comment (Monkeyanator) [reply operator/cmd/mesh/install.go:189]: Fixed.

**Comment by Monkeyanator** [comment]
@costinm @howardjohn believe I've addressed most of the feedback, PTAL when you have the chance.

**Comment by howardjohn** [comment]
It seems like `istiod-default-validator` is not patched by any istiod revision. Works for most cases, but if cabundle changes it will explode

File: pkg/webhooks/validation/controller/controller.go
Comment (howardjohn) [inline pkg/webhooks/validation/controller/controller.go:323]: I would prefer `c.webhookInformer.GetStore().List()`. Its cached already, so no call to api-server like here, and already has filter logic for `-l istio.io/rev=<rev>`

Comment (howardjohn) [reply pkg/webhooks/validation/controller/controller.go:323]: Or if not, we should probably at least filter?

**Review by howardjohn — APPROVED** [review]

File: pkg/webhooks/validation/controller/controller.go
Comment (Monkeyanator) [reply pkg/webhooks/validation/controller/controller.go]: Makes sense, fixed

**Comment by jacob-delgado** [comment]
What are the steps to migrate users from the older chart to the new ones? This is starting to look pretty good. The helm changes look fine, but some clarity on steps would be useful.

**Comment by Monkeyanator** [comment]
> What are the steps to migrate users from the older chart to the new ones?

The only action required is that Helm users will have to add `--defaultRevision default` (or whatever revision) when installing the new base charts. This will (1) remove the old validator and (2) create the new default validator `istio-default-validator`.

File: pkg/webhooks/validation/controller/controller.go
Comment (Monkeyanator) [inline pkg/webhooks/validation/controller/controller.go:179]: Deleted this since I don't believe we need to reconcile on deletion and it caused race conditions when switching revision tag on default validator

**Review by jacob-delgado — APPROVED** [review]
Helm changes lgtm and doc is approved

**Review by esnible — COMMENTED** [review]
I like the direction.

Would it be reasonable to also modify `revision list` to add columns for both mutating and validating webhooks?

Would it be reasonable to modify `revision describe <tag>` to add section for validating webhook?

**Comment by Monkeyanator** [comment]
> Would it be reasonable to also modify revision list to add columns for both mutating and validating webhooks?

@esnible I think we should avoid a adding column for validating webhook to `istioctl tag list` since we only expect this to ever happen with the `default` tag. Might also leak too much implementation detail about how tags are implemented in the command output. 

IMO we just need to make sure that the meaning of the `default` tag is well documented (both in the command help text and actual docs).

**Comment by Monkeyanator** [comment]
@istio/wg-docs-maintainers @istio/wg-user-experience-maintainers PTAL when you have the chance, thanks!

**Review by ericvn — APPROVED** [review]

**Review by therealmitchconnors — APPROVED** [review]

**Comment by ericvn** [comment]
As noted, this PR needs some istio.io changes and moving the istio.io tests to after this commit has caused some failures. See  https://github.com/istio/istio.io/pull/10208

**Pull request merged**
